sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/ProcessBundleBenchmark.java

```diff
@@ -283,4 +300,207 @@ public void testLargeBundle(TrivialTransform trivialTransform) throws Exception
+      }
+      return CompletableFuture.completedFuture(response);
+    }
+  }
+
+  @Benchmark
+  @Threads(16) // Use several threads since we expect contention during bundle processing.
+  public void testStateWithoutCaching(StatefulTransform statefulTransform) throws Exception {
+    testState(statefulTransform, statefulTransform.nonCachingStateRequestHandler);
+  }
+
+  @Benchmark
+  @Threads(16) // Use several threads since we expect contention during bundle processing.
+  public void testStateWithCaching(StatefulTransform statefulTransform) throws Exception {
+    testState(statefulTransform, statefulTransform.cachingStateRequestHandler);
+  }
+
+  private static void testState(
+      StatefulTransform statefulTransform, StateRequestHandler stateRequestHandler)
+      throws Exception {
+    Map<String, ? super Coder<WindowedValue<?>>> remoteOutputCoders =
+        statefulTransform.descriptor.getRemoteOutputCoders();
+    Map<String, RemoteOutputReceiver<?>> outputReceivers = new HashMap<>();
+    AtomicInteger outputValuesCount = new AtomicInteger();
+    for (Entry<String, ? super Coder<WindowedValue<?>>> remoteOutputCoder :
+        remoteOutputCoders.entrySet()) {
+      outputReceivers.put(
+          remoteOutputCoder.getKey(),
+          RemoteOutputReceiver.of(
+              (Coder) remoteOutputCoder.getValue(),
+              (FnDataReceiver<? super WindowedValue<?>>)
+                  (WindowedValue<?> value) -> outputValuesCount.incrementAndGet()));
+    }
+    String key = padLeftZeros(Long.toHexString(Thread.currentThread().getId()), 16);
+    try (RemoteBundle bundle =
+        statefulTransform.processor.newBundle(
+            outputReceivers, stateRequestHandler, BundleProgressHandler.ignored())) {
+      Iterables.getOnlyElement(bundle.getInputReceivers().values())
+          .accept(valueInGlobalWindow(KV.of(key, "zero")));
+    }
+    try (RemoteBundle bundle =
+        statefulTransform.processor.newBundle(
+            outputReceivers, stateRequestHandler, BundleProgressHandler.ignored())) {
+      Iterables.getOnlyElement(bundle.getInputReceivers().values())
+          .accept(valueInGlobalWindow(KV.of(key, "one")));
+    }
+    try (RemoteBundle bundle =
+        statefulTransform.processor.newBundle(
+            outputReceivers, stateRequestHandler, BundleProgressHandler.ignored())) {
+      Iterables.getOnlyElement(bundle.getInputReceivers().values())
+          .accept(valueInGlobalWindow(KV.of(key, "two")));
+    }
+    try (RemoteBundle bundle =
+        statefulTransform.processor.newBundle(
+            outputReceivers, stateRequestHandler, BundleProgressHandler.ignored())) {
+      Iterables.getOnlyElement(bundle.getInputReceivers().values())
+          .accept(valueInGlobalWindow(KV.of(key, "flush")));
+    }
+    assertEquals(3, outputValuesCount.getAndSet(0));
+  }
+
+  private static String padLeftZeros(String inputString, int length) {
```

Comment:
Can we use a library to do this? e.g. https://guava.dev/releases/19.0/api/docs/com/google/common/base/Strings.html#padStart(java.lang.String,%20int,%20char)